packages/devtools_app/lib/src/screens/memory/panes/chart/controller/chart_connection.dart

```diff
@@ -38,28 +40,59 @@ class ChartConnection extends DisposableController
   Timer? _pollingTimer;
   bool _connected = false;
 
-  late final isDeviceAndroid =
-      serviceConnection.serviceManager.vm?.operatingSystem == 'android';
+  /// If true, the connection to application was lost.
+  bool _disconnected = false;
+
+  late final isDeviceAndroid = _isDevToolsCurrentlyConnected()
+      ? serviceConnection.serviceManager.vm?.operatingSystem == 'android'
+      : false;
+
+  /// True if DevTools is in connected mode and the connection to the app is alive.
+  bool _isDevToolsCurrentlyConnected() =>
+      !offlineDataController.showingOfflineData.value &&
+      serviceConnection.serviceManager.connectedState.value.connected &&
+      serviceConnection.serviceManager.connectedApp != null;
```

Comment:
I think you can drop this check. Checking `connectedState.value.connected` should be enough

Reply:
It is not enough. I do not know what is going on, but these values are not always in sync during loading offline data and/or app disconnect.
Added comment.

Reply:
Okay I've put together a suggestion of what to do with this class. As is, this is starting to feel a bit messy to have to call everything in a `_runSafely` wrapper. Other code in DevTools does not have to do this by making sure that the `connectedState` notifier is being listened to. If you apply these changes, you will not see an error on app disconnect and the polling timer will be handled properly:

https://github.com/flutter/devtools/compare/master...kenzieschmoll:polina-code-review?expand=1

Reply:
Yes, what you suggest may pass tests, but I have concerns about reliability.
The code that requires connection is async. That means the connection state can change in an async gap. That means checking connection state before operation will not protect from failures.

And, the actual connection may become unavailable before DevTools detected it and updated connectedState.

Thoughts?

Reply:
Updated as discussed offline.